ACLs: Support external servers

ishustava opened the pull request:
## Problem
Currently, we support ACL bootstrapping only if the servers are also deployed on Kubernetes. If you're running servers elsewhere, we currently recommend bootstrapping ACLs yourself and setting up the correct policies and tokens for the k8s components. You could then set them as values, e.g. `syncCatalog.aclSyncToken` or `connectInject.overrideAuthMethodName` and `connectInject.aclInjectToken`. 

However, figuring out the right policies is hard, and creating them yourself is still a manual process. Even though servers are running externally, our knowledge about the policies and tokens you need doesn't change since we still consume them from the Helm chart. The workaround right now is to create a Kubernetes secret with the bootstrap token named something that the Helm chart expects. This will trick the job into thinking that the bootstrapping was already done and it will continue to create tokens and policies for the rest of the components.

See #413 for a bit more details.

## Proposal
This PR proposes to use the new `externalServers` configuration to enable the server-acl-init job to talk to external servers. 

It also provides two ways of "initializing" ACLs for external servers:
1. You can run bootstrap ACLs yourself via the `acl/bootstrap` API and then provide the bootstrap token by setting `global.acls.bootstrapToken` secret. The Helm config for this use case could look like this:
    ```yaml
    global:
      enabled: false
      acls:
        manageSystemACLs: true
        bootstrapToken:
          secretName: consul-bootstrap-acl-token
          secretKey: token
    externalServers:
      enabled: true
    client:
      enabled: true
      join: [<retry_join address(es)>]
    connectInject:
      enabled: true
      overrideAuthMethodHost: <external address of the kube api server>
    ...
    ```
1.  You can let the server-acl-init job to call the bootstrapping API. It will behave in the same way as when the servers are running on k8s and will save the bootstrap token as a Kubernetes secret. In this case, the Helm config will look like this:
    ```yaml
    global:
      enabled: false
      acls:
        manageSystemACLs: true
    externalServers:
      enabled: true
    client:
      enabled: true
      join: [<retry_join address(es)>]
    connectInject:
      enabled: true
      overrideAuthMethodHost: <external address of the kube api server>
    ...
    ```

### Testing
This PR can be tested  on AKS with the Helm configs above and the `ishustava/consul-k8s-dev:04-16-2020-ae132bc` image.

Requires #401 to be merged first.

pratiklotia commented:
Have a question @ishustava 

Great suggestion for the new feature. My question is, after adding this capability, how can we customize which client node gets what type of policy (read/write) and which service gets what policy. Looks like the additional configuration in the helm chart here, will ask for a 'standard' policy to be set on the consul server? For example, using this - all client nodes (in the form of a k8s node) will get the same policy, correct? Or is there room to customize policy per client node?

ishustava commented:
Hey @pratiklotia, thanks for your feedback. Correct, at the moment, all client pods are given the same token even if servers are deployed on k8s too. 

This is a good point though and is something that has come up before. I think we should look into giving clients a unique token and making their policies more restrictive, but I think this is out-of-scope for this feature. I've filed #425 to track this separately.

adilyse commented on test/unit/server-acl-init-job.bats:
```diff
@@ -800,3 +800,201 @@ load _helpers
+  [ "${actual}" = "true" ]
+}
+
+@test "serverACLInit/Job: can override externalServers.https.port" {
```
It's good to also test the default value here. That makes it very clear anyone that they're changing defaults and it might be a breaking change.

adilyse commented on test/unit/server-acl-init-job.bats:
```diff
@@ -800,3 +800,201 @@ load _helpers
+  [ "${actual}" = "false" ]
+}
+
+@test "serverACLInit/Job: doesn't set the CA cert if TLS is enabled and externalServers.https.useSystemRoots is true" {
```
The other variations of this should be tested too, to make sure that the CA cert _is_ set if these values aren't enabled.

adilyse commented on test/unit/server-acl-init-job.bats:
```diff
@@ -800,3 +800,201 @@ load _helpers
+  yq '.spec.template.spec.containers[0].volumeMounts | map(select(.name == "bootstrap-token")) | length == 1' | tee /dev/stderr)
+  [ "${actual}" = "true" ]
+}
+
```
There should be tests for the case where both the bootstrap token and replication token are set.

adilyse commented on test/unit/server-acl-init-job.bats:
```diff
@@ -800,3 +800,201 @@ load _helpers
+#--------------------------------------------------------------------
+# connectInject.authMethodConfig
+
+@test "serverACLInit/Job: can provide custom auth method configuration" {
```
This also should have a default test case, to make sure they're not set if the flags aren't defined.

ishustava commented on values.yaml:
```diff
@@ -307,10 +316,10 @@ server:
+# This configuration is recommended if setting global.tls.enableAutoEncrypt to true (requires consul-k8s >= 0.13.0)
+# or global.acls.manageSystemACLs to true (requires consul-k8s >= 0.14.0) and running
+# Consul servers outside of the current Kubernetes cluster.
 externalServers:
```
We talked about this rename [here](https://github.com/hashicorp/consul-helm/pull/375#discussion_r398825308). Let me know if this reasoning makes sense to you.

ishustava commented on templates/server-acl-init-job.yaml:
```diff
@@ -120,7 +150,9 @@ spec:
                 {{- end }}
-                {{- if (and .Values.global.acls.replicationToken.secretName .Values.global.acls.replicationToken.secretKey) }}
+                {{- if (and .Values.global.acls.bootstrapToken.secretName .Values.global.acls.bootstrapToken.secretKey) }}
+                -bootstrap-token-file=/consul/acl/tokens/bootstrap-token \
```
Yep, they are. The reason is that if the replication token is set, then that token is used to do acl-init so there's no need for the bootstrap token. Yeah, I think adding a note would be good.

ishustava commented on templates/server-acl-init-job.yaml:
```diff
@@ -101,6 +129,8 @@ spec:
                 {{- end }}
                 {{- if .Values.connectInject.enabled }}
                 -create-inject-auth-method=true \
+                -inject-auth-method-host={{ .Values.connectInject.authMethodConfig.host }} \
```
Yeah we should. I was hoping to get away with less yaml templating logic, but forgot that it will look weird with these flags set to empty strings when you'll try to describe the pod. I'll add the ifs. They don't need to both be set because if you haven't set them we'll just fall back to using the default (either the kube api service or the CA included with the service account token).

ishustava commented on values.yaml:
```diff
@@ -781,6 +790,21 @@ connectInject:
+  # This is useful when Consul servers are deployed outside of
+  # the Kubernetes cluster. Requires consul-k8s >= 0.14.0.
+  authMethodConfig:
+    # host corresponds to the Consul's Host configuration parameter
+    # ref: https://www.consul.io/docs/acl/auth-methods/kubernetes.html#host.
```
It has to an addess that the consul server can reach. Good call, I'll add a clarification.

ishustava commented on values.yaml:
```diff
@@ -781,6 +790,21 @@ connectInject:
+
+    # caCert is the base64-encoded PEM-encoded CA certificate string
+    # for this Kubernetes cluster.
+    caCert: ""
```
Yeah, we can access it from the service account. I'm adding this config value in case you have a setup where the CA included in the service account would not be the same as the CA for the kube api host. 

You could provide only the host, I'll make it more clear in the comment for this option.

adilyse replied:
How common is it for someone to have a different CA in the service account and the kube api host?

ishustava replied:
Thinking about this a bit more, I don't think it makes too much sense to provide a custom CA. I'd like to punt on this and wait for someone to ask to make it configurable and, for the time being, only allow the host to be configurable.

ishustava commented on values.yaml:
```diff
@@ -307,10 +316,10 @@ server:
+# or global.acls.manageSystemACLs to true (requires consul-k8s >= 0.14.0) and running
+# Consul servers outside of the current Kubernetes cluster.
 externalServers:
   enabled: false
```
Yep, will add a comment. The comment for the `externalServers.https.address` property already mentions the point about `client.join` property. Do think that it should be mentioned again here?

adilyse replied:
I think so? Particularly since that setting isn't in this section, it's worth being very explicit about the crossover.

ishustava replied:
Added clarification.

ishustava commented on values.yaml:
```diff
@@ -781,6 +790,21 @@ connectInject:
   # auth method for Connect inject, set this to the name of your auth method.
   overrideAuthMethodName: ""
 
+  # Custom configuration for the Kubernetes auth method to be used when
+  # global.acls.manageSystemACLs is set to true. Please see
+  # https://www.consul.io/docs/acl/auth-methods/kubernetes.html.
```
Currently, the tests for these values don't depend on external servers being set. The job also doesn't depend on the value of external servers. It should work either way, although I don't see why anyone would set these values for anything other than external servers. 

I've chosen to put it under connectInject together with other ACL-related config values for connect injector, e.g. `aclBindingRuleSelector`. I think thematically it makes sense.

Are you suggesting that it should be moved to the externalServers section? Why do you think it's better?

ishustava commented:
@adilyse Thanks so much for the review and organizing your thoughts so clearly! I've updated the PR with your suggestions. 

The main change that came out of that was the configuration change. I think allowing configuring a custom CA cert for the auth method will be pretty uncommon, and so I'd rather punt on adding that configuration. As a result, the `authMethodConfig` became just one property `overrideAuthMethodHost` to match `overrideAuthMethodName`. 

With regards to the placement of this config property, I chose the `connectInject` section because it seemed more specific and because it already had other acl-related properties. In general, this property applies when all three are true: ACLs are enabled, external servers are enabled, and connect inject is enabled. I'm not sure how to choose the best section out of these three, and so I went with `connectInject` because that seemed the most consistent with other existing values, like `aclBindingRuleSelector`. I'm definitely open to suggestions.

---

Below are my responses to your other questions:

> Should we warn folks using external servers if their client.join and externalServers.https.address values don't match?

No, because the purpose of this property is to be able to overwrite the client.join address used by default.

> Should we warn folks if externalServers.enabled=true and servers are enabled?

Yep, I think that's a good idea. I'll add a failure to the templates and tests. I've also added a warning to the comment for the `externalServers.enabled` property.

> https: Is this entire block optional since we fall back to the client.join setup?

Yes.

> https:address: It would be helpful to mention that this can't include a port.

The comment already lists all the values that this property _can_ take, i.e. DNS name, IP or cloud auto-join string. This implies that it should not include the port and I feel like mentioning it would be unnecessary.

> https.port: Would there ever be a case where I've set my client.join property, but the port is pulled from here? Could that cause problems?

I'm not sure what you mean; maybe you have an example in mind? Your `client.join` property doesn't include the port, so you'd need to provide it somehow, otherwise, there's no easy way for us to know which HTTP(s) port to use.

> tlsServerName and useSystemRoots: Are these only useful settings when using a custom CA?

`tlsServerName` is useful when the address you're connecting to is not included in your server certificate. Then you can set tlsServerName to something that is included in the certificate so that hostname verification part of the TLS handshake won't fail.
`useSystemRoots` is useful when your CA is not a custom CA but a known CA that is typically included in your OS system roots, for example, Let's Encrypt CA. In this case, we don't need to provide a CA for TLS connections.

> Should externalServers.https.address be an array? It mirrors client.join which is an array.

We don't need this to be an array because we only need one server to talk to, doesn't matter which one.

> Any logic that references both server and externalServers needs to do an exclusion test by setting server.enabled to false since servers are enabled by default in the Helm chart.

Good call! I've added missing tests.

adilyse commented on values.yaml:
```diff
@@ -777,10 +792,17 @@ connectInject:
+  # This address must to be reachable from the Consul servers.
+  # Please see https://www.consul.io/docs/acl/auth-methods/kubernetes.html.
+  # Requires consul-k8s >= 0.14.0.
+  overrideAuthMethodHost: ""
```
I still think this makes more sense to be under the `externalServers` section. You _can_ set it when not using external servers, but it's unlikely that very many people will need. On the other hand, it could be quite common for folks running external servers to need to make use of this value. It could be easy to miss it though, since it's not grouped with those settings.

adilyse commented on templates/server-acl-init-job.yaml:
```diff
@@ -83,16 +96,32 @@ spec:
+                {{- if not (and .Values.externalServers.enabled .Values.externalServers.https.useSystemRoots) }}
                 -consul-ca-cert=/consul/tls/ca/tls.crt \
+                {{- end }}
+                {{- if not .Values.externalServers.enabled }}
                 -server-port=8501 \
                 {{- end }}
```
This might make sense to be set near the `-server-address` flag.

ishustava commented:
If somebody has provided a client.join with a port, then I think we can make the server-acl-init command ignore the port if it's provided to the `server-address` flag.

I'm not super concerned with port being overwritten because `client.join` will contain the serf port, while we need an HTTP port, so you would need to provide it anyway.

ishustava commented on values.yaml:
```diff
@@ -187,11 +186,14 @@ global:
     # replicationToken references a secret containing the replication ACL token.
     # This token will be used by secondary datacenters to perform ACL replication
     # and create ACL tokens and policies.
+    # # This value is ignored if bootstrapToken is also set.
```
ty!
